src/tools/AutomanProxy.sol

```diff
@@ -0,0 +1,313 @@
+
+    function rebalance(
+        INPM.MintParams memory params,
+        uint256 tokenId,
+        uint256 feePips,
+        bytes calldata swapData
+    ) external returns (uint256 newTokenId, uint128 liquidity, uint256 amount0, uint256 amount1);
+
+    function removeLiquidity(
+        INPM.DecreaseLiquidityParams memory params,
+        uint256 feePips
+    ) external returns (uint256 amount0, uint256 amount1);
+
+    function removeLiquidity(
+        INPM.DecreaseLiquidityParams memory params,
+        uint256 feePips,
+        uint256 permitDeadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external returns (uint256 amount0, uint256 amount1);
+
+    function removeLiquiditySingle(
+        INPM.DecreaseLiquidityParams memory params,
+        bool zeroForOne,
+        uint256 feePips,
+        bytes calldata swapData
+    ) external returns (uint256 amount);
+
+    function removeLiquiditySingle(
+        INPM.DecreaseLiquidityParams memory params,
+        bool zeroForOne,
+        uint256 feePips,
+        bytes calldata swapData,
+        uint256 permitDeadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external returns (uint256 amount);
+
+    function reinvest(
+        INPM.IncreaseLiquidityParams memory params,
+        uint256 feePips,
+        bytes calldata swapData
+    ) external returns (uint128 liquidity, uint256 amount0, uint256 amount1);
+
+    function reinvest(
+        INPM.IncreaseLiquidityParams memory params,
+        uint256 feePips,
+        bytes calldata swapData,
+        uint256 permitDeadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external returns (uint128 liquidity, uint256 amount0, uint256 amount1);
+}
+
+contract AutomanRelayerProxy is Ownable, Automan {
+    mapping(address => mapping(address => bool)) public allowance;
+    INPM public npm;
+    Automan public automan;
+    constructor(INPM _npm, Automan _automan) Ownable(msg.sender) {
+        npm = _npm;
+        automan = _automan;
+    }
+
+    function setAllowance(address[] calldata relayers, address[] calldata owners, bool value) external onlyOwner {
+        uint lrelayer = relayers.length;
+        uint lowners = owners.length;
+        for (uint i = 0; i < lrelayer; i++) {
+            for (uint j = 0; j < lowners; j++) {
+                allowance[relayers[i]][owners[j]] = value;
+            }
+        }
+    }
+
+    function updateNpm(INPM _npm) external onlyOwner {
+        npm = _npm;
+    }
+
+    function rebalance(
+        INPM.MintParams memory params,
+        uint256 tokenId,
+        uint256 feePips,
+        bytes calldata swapData,
+        uint256 permitDeadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external returns (uint256 newTokenId, uint128 liquidity, uint256 amount0, uint256 amount1) {
+        address owner = npm.ownerOf(tokenId);
```

Comment:
Make the following logic into a modifier (similar to how we use `onlyOwner`):

```
address owner = npm.ownerOf(tokenId);
require(allowance[msg.sender][owner], "not allow relayer");
```

Reply:
done